### PR TITLE
WT-7636 Fix leaked cursors by implementing scoped cursor and session types

### PR DESCRIPTION
**Pull request by tetsuo-cpp**
Implement some RAII types to manage the lifetimes of cursors and sessions in our system. I've introduced two classes in particular: `scoped_cursor` and `scoped_session`.

They automatically close themselves when they fall out of scope and their destructor is invoked.

### CODE REVIEW

**Comment by tetsuo-cpp** [comment]
The good news is that when we turn on workload validation, we're no longer getting cache stuck. However, the connection close checkpoint does seem to take an incredibly long time so there's still something not right.

**Review by etienneptl — APPROVED** [review]
LGTM

File: test/cppsuite/test_harness/util/scoped_types.h [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+namespace test_harness {
+
+class scoped_cursor {
+    public:
+    scoped_cursor() = default;
```
Comment (luke-pearson) [inline test/cppsuite/test_harness/util/scoped_types.h:33]: If a caller was to use this constructor they'd be left with a non initialised cursor right?

File: test/cppsuite/test_harness/util/scoped_types.h [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    scoped_session(scoped_session &&other)
+    {
+        std::swap(_session, other._session);
+    }
+
+    scoped_session &
+    operator=(scoped_session &&other)
+    {
+        scoped_session tmp(std::move(other));
+        std::swap(_session, tmp._session);
```
Comment (luke-pearson) [inline test/cppsuite/test_harness/util/scoped_types.h:119]: Can you add some comments around the more C++y things that you've done here so its clear what is happening to non c++ readers?

File: test/cppsuite/test_harness/util/scoped_types.h (hunk 1 above)
Comment (tetsuo-cpp) [reply test/cppsuite/test_harness/util/scoped_types.h:33]: Yep, that's right. It'll just be `nullptr` and can move assign to it.

File: test/cppsuite/test_harness/util/scoped_types.h [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    /* Moving is ok but copying is not. */
+    scoped_cursor(scoped_cursor &&other)
+    {
+        std::swap(_cursor, other._cursor);
+    }
+
+    /*
+     * Implement move assignment by using move constructing a temporary and swapping its internals
```
Comment (luke-pearson) [inline test/cppsuite/test_harness/util/scoped_types.h:52]: I think the grammar here is odd, namely: "by using move contructing a temporary"

**Review by luke-pearson — APPROVED** [review]
LGTM with one comment fix.

File: test/cppsuite/test_harness/util/scoped_types.h (hunk 2 above)
Comment (tetsuo-cpp) [reply test/cppsuite/test_harness/util/scoped_types.h:135]: 2190aae

File: test/cppsuite/test_harness/util/scoped_types.h (hunk 3 above)
Comment (tetsuo-cpp) [reply test/cppsuite/test_harness/util/scoped_types.h]: 1d1259c

**Pull request merged**